encryptor/queryDataEncryptor.go

```diff
@@ -313,6 +316,187 @@ func (encryptor *QueryDataEncryptor) OnQuery(query base.OnQueryObject) (base.OnQ
 	return query, false, nil
 }
 
+// OnBind process bound values for prepared statement based on TableSchemaStore.
+func (encryptor *QueryDataEncryptor) OnBind(statement sqlparser.Statement, values []base.BoundValue) ([]base.BoundValue, bool, error) {
+	newValues := values
+	changed := false
+	var err error
+	switch statement := statement.(type) {
+	case *sqlparser.Insert:
+		newValues, changed, err = encryptor.encryptInsertValues(statement, values)
+	case *sqlparser.Update:
+		newValues, changed, err = encryptor.encryptUpdateValues(statement, values)
+	}
+	if err != nil {
+		return values, false, err
+	}
+	return newValues, changed, nil
+}
+
+func (encryptor *QueryDataEncryptor) encryptInsertValues(insert *sqlparser.Insert, values []base.BoundValue) ([]base.BoundValue, bool, error) {
+	logrus.Debugln("QueryDataEncryptor.encryptInsertValues")
+	tableName := insert.Table.Name
+	// Look for the schema of the table where the INSERT happens.
+	// If we don't have a schema then we don't know what to encrypt, so do nothing.
+	schema := encryptor.schemaStore.GetTableSchema(tableName.String())
+	if schema == nil {
+		logrus.WithField("table", tableName).Debugln("No encryption schema")
+		return values, false, nil
+	}
+
+	// Gather column names from the INSERT query. If there are no columns in the query,
+	// expect a complete list of colums to be available in the schema.
+	var columns []string
+	if len(insert.Columns) > 0 {
+		columns = make([]string, len(insert.Columns))
+		for i, column := range insert.Columns {
+			columns[i] = column.String()
+		}
+	} else if cols := schema.Columns(); len(cols) > 0 {
+		columns = cols
+	}
+
+	changed := false
+	newValues := make([]base.BoundValue, len(values))
```

Comment:
can we move it inside `if` statement to not allocate and copy values when we don't need it? plus we can move `if changed` check with `return` too.

Reply:
> can we move it inside if statement to not allocate and copy values when we don't need it?

I'll see what I can do. At worst, we can check the `changed` value and allocate on the first change.

Reply:
I've cleaned up the code, extracted it into a utility method shared by INSERT and UPDATE implementations. Now it won't allocate a new slice if none of the values need to be encrypted. I've also fixed an issue in the original code while at it.